pkg/issuer/venafi/client/venaficlient.go

```diff
@@ -70,15 +82,90 @@ type connector interface {
 	RenewCertificate(req *certificate.RenewalRequest) (requestID string, err error)
 }
 
-func New(namespace string, secretsLister corelisters.SecretLister, issuer cmapi.GenericIssuer) (Interface, error) {
-	cfg, err := configForIssuer(issuer, secretsLister, namespace)
+func secretNameForIssuer(venCfg *cmapi.VenafiIssuer) string {
+	switch {
+	case venCfg.TPP != nil:
+		return venCfg.TPP.CredentialsRef.Name
+	case venCfg.Cloud != nil:
+		return venCfg.Cloud.APITokenSecretRef.Name
+	}
+	return ""
+}
+
+func New(namespace string, secretsLister corelisters.SecretLister, issuer cmapi.GenericIssuer, coreClient coreclientset.CoreV1Interface) (Interface, error) {
+	venCfg := issuer.GetSpec().Venafi
+	secret, err := secretsLister.Secrets(namespace).Get(secretNameForIssuer(venCfg))
 	if err != nil {
 		return nil, err
 	}
-
-	vcertClient, err := vcert.NewClient(cfg)
+	secret = secret.DeepCopy()
+	cfg, err := configForIssuer(issuer, secret)
 	if err != nil {
-		return nil, fmt.Errorf("error creating Venafi client: %s", err.Error())
+		return nil, err
+	}
+
+	var vcertClient endpoint.Connector
+
+	switch cfg.ConnectorType {
+	case endpoint.ConnectorTypeTPP:
+		var connectionTrustBundle *x509.CertPool
+		if cfg.ConnectionTrust != "" {
+			connectionTrustBundle = x509.NewCertPool()
+			if !connectionTrustBundle.AppendCertsFromPEM([]byte(cfg.ConnectionTrust)) {
+				return nil, fmt.Errorf("%w: failed to parse PEM trust bundle", verror.UserDataError)
+			}
+		}
+		tppClient, err := tpp.NewConnector(cfg.BaseUrl, cfg.Zone, cfg.LogVerbose, connectionTrustBundle)
+		if err != nil {
+			return nil, fmt.Errorf("error creating TPP connector: %v", err)
+		}
+
+		if cfg.Credentials.RefreshToken != "" {
+			oauth, err := tppClient.RefreshAccessToken(cfg.Credentials)
+			if err != nil {
+				return nil, fmt.Errorf("error refreshing access token: %v", err)
+			}
+			oauthBytes, err := json.Marshal(oauth)
+			if err != nil {
+				return nil, fmt.Errorf("error marshalling oauth tokens: %v", err)
+			}
+			originalBytes, err := json.Marshal(secret)
+			if err != nil {
+				return nil, fmt.Errorf("error marshalling original secret: %v", err)
+			}
+			secret.Data[oauthKey] = oauthBytes
+			newBytes, err := json.Marshal(secret)
+			if err != nil {
+				return nil, fmt.Errorf("error marshalling updated secret: %v", err)
+			}
+			patchBytes, err := strategicpatch.CreateTwoWayMergePatch(originalBytes, newBytes, corev1.Secret{})
+			if err != nil {
+				return nil, fmt.Errorf("error creating patch: %v", err)
+			}
+
+			_, err = coreClient.Secrets(namespace).Patch(
```

Comment:
Is there a reason for using a Patch here? Do we anticipate any other actors concurrently modifying this Secret resource?